packages/material-ui/src/Tooltip/Tooltip.js

```diff
@@ -72,120 +73,163 @@ export const styles = theme => ({
   },
 });
 
-class Tooltip extends React.Component {
-  ignoreNonTouchEvents = false;
-
-  constructor(props) {
-    super();
-    this.isControlled = props.open != null;
-    this.state = {
-      open: null,
-    };
-
-    if (!this.isControlled) {
-      // not controlled, use internal state
-      this.state.open = false;
+function Tooltip(props) {
+  const {
+    children,
+    classes,
+    disableFocusListener,
+    disableHoverListener,
+    disableTouchListener,
+    enterDelay,
+    enterTouchDelay,
+    id,
+    interactive,
+    leaveDelay,
+    leaveTouchDelay,
+    onClose,
+    onOpen,
+    open: openProp,
+    placement,
+    PopperProps,
+    theme,
+    title,
+    TransitionComponent,
+    TransitionProps,
+    ...other
+  } = props;
+
+  const [openState, setOpenState] = React.useState(false);
+  const [, forceUpdate] = React.useState(0);
+  const [childNode, setChildNode] = React.useState();
+  const ignoreNonTouchEvents = React.useRef(false);
+  const { current: isControlled } = React.useRef(props.open != null);
+  const defaultId = React.useRef();
+  const closeTimer = React.useRef();
+  const enterTimer = React.useRef();
+  const leaveTimer = React.useRef();
+  const touchTimer = React.useRef();
+  // can be removed once we drop support for non ref forwarding class components
+  const handleOwnRef = React.useCallback(ref => {
+    setChildNode(ReactDOM.findDOMNode(ref));
+  }, []);
+  const handleRef = useForkRef(children.ref, handleOwnRef);
+
+  React.useEffect(() => {
+    if (childNode) {
```

Comment:
We are mixing the `if (x) { warning(message) }` and `warning(y, message)` pattern. We should only use one strategy. React is introducing new helpers: https://github.com/facebook/react/pull/15170. We can start to migrate to `if(x) { warn(message) }` if you prefer or stick to the existing convention.

Reply:
I did want to refactor this but it's my kryptonite ... Boolean expressions 😅